docs/hub/datasets-spark.md

```diff
@@ -0,0 +1,441 @@
+            revision = resolved_path.revision,
+            repo_type=resolved_path.repo_type,
+            operations=[
+                CommitOperationCopy(src_path_in_repo=path, path_in_repo=new_path)
+                for path, new_path in new_paths.items()
+            ] + [
+                CommitOperationDelete(path_in_repo=path)
+                for path in paths
+            ],
+            commit_message="Rename files" + (f" (part {i:05d}-of-{num_commits:05d})" if num_commits > 1 else "")
+        )
+        yield pa.record_batch(
+                {
+                    "path": list(new_paths.values()),
+                },
+                schema=pa.schema({"path": pa.string()})
+            )
+
+
+def write_parquet(df: DataFrame, path: str, **kwargs) -> None:
+    """
+    Write Parquet files to Hugging Face using PyArrow.
+
+    It uploads Parquet files in a distributed manner in three steps:
+
+    1. Preupload the Parquet files in parallel in a distributed banner
+    2. Commit the preuploaded files
+    3. Rename the files to their final names
+
+    Authenticate using `huggingface-cli login` or passing a token
+    using the `storage_options` argument: `storage_options={"token": "hf_xxx"}`
+
+    Parameters
+    ----------
+    path : str
+        Path of the file or directory. Prefix with a protocol like `hf://` to read from Hugging Face.
+        It writes Parquet files in the form "part-xxxxx.parquet", or to a single file if `path ends with ".parquet".
+
+    **kwargs
+        Any additional kwargs are passed to pyarrow.parquet.ParquetWriter.
+
+    Returns
+    -------
+    DataFrame
+        DataFrame based on parquet file.
+
+    Examples
+    --------
+    >>> spark.createDataFrame(pd.DataFrame({"foo": range(5), "bar": range(5, 10)}))
+    >>> # Save to one file
+    >>> write_parquet(df, "hf://datasets/username/dataset/data.parquet")
+    >>> # OR save to a directory (possibly in many files)
+    >>> write_parquet(df, "hf://datasets/username/dataset")
+    """
+    filesystem: HfFileSystem = kwargs.pop("filesystem", HfFileSystem(**kwargs.pop("storage_options", {})))
+    if path.endswith(".parquet") or path.endswith(".pq"):
+        df.coalesce(1).mapInArrow(
+            partial(_write_single, path=path, schema=to_arrow_schema(df.schema), filesystem=filesystem, **kwargs),
+            from_arrow_schema(pa.schema({"path": pa.binary()})),
+        ).collect()
+    else:
+        df.mapInArrow(
+            partial(_preupload, path=path, schema=to_arrow_schema(df.schema), filesystem=filesystem, **kwargs),
+            from_arrow_schema(pa.schema({"addition": pa.binary()})),
+        ).coalesce(1).mapInArrow(
+            partial(_commit, path=path, filesystem=filesystem),
+            from_arrow_schema(pa.schema({"path": pa.string()})),
+        ).coalesce(1).mapInArrow(
+            partial(_rename, path=path, filesystem=filesystem),
+            from_arrow_schema(pa.schema({"path": pa.string()})),
```

Comment:
What the purpose of committing and then renaming? I feel that the code would be more readable if we only preupload+commit (which is already complex logic in itself)

Reply:
Because when uploading I don't have the shard_idx information ^^'

Is it possible to do the CommitOperationAdd with different path_in_repo than the ones used for preupload_lfs ? This way now need for the rename step.

Reply:
Or maybe @anton-l if you know a way I can get the id of the partition passed in mapInArrow that I can use to get the shard_idx

Reply:
Yes you can define whatever `path_in_repo` you want in the preupload step and change it afterwards before the `commit` step. It's hacky but the whole script seems to be for power users so :man_shrugging: In fact in the preupload step the `path_in_repo` is only used to determine if the file should be uploaded as LFS or not (rule is either "filesize > threshold" or "file extension in .gitattributes").

Reply:
I just simplified the code (significantly) using your trick, let me know what you think :)